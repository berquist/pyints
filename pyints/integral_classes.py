"""Generic classes that define the properties of each integral type
without knowing an interface to each program."""


class Integral:
    pass


class Integral1e2c(Integral):
    pass


class Integral2e4c(Integral):
    pass


class Overlap(Integral1e2c):
    pass


class Kinetic(Integral1e2c):
    pass


class Potential(Integral1e2c):
    pass


class FermiContact(Integral1e2c):
    pass


FC = FermiContact


class SpinDipole(Integral1e2c):
    pass


SD = SpinDipole


class DipoleLength(Integral1e2c):
    pass


Dipole = DipoleLength


class SecondMoment(Integral1e2c):
    pass


class Quadrupole(Integral1e2c):
    pass


class TracelessQuadrupole(Integral1e2c):
    pass


class MultipoleLength(Integral1e2c):
    pass


Multipole = MultipoleLength


class DipoleVelocity(Integral1e2c):
    pass


Nabla = DipoleVelocity


class AngularMomentum(Integral1e2c):
    pass


class OverlapDer1(Integral1e2c):
    pass


class KineticDer1(Integral1e2c):
    pass


class PotentialDer1(Integral1e2c):
    pass


class Repulsion(Integral2e4c):
    pass


class SpinOrbitExact(Integral2e4c):
    pass
