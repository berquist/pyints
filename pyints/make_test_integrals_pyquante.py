from __future__ import print_function

import numpy as np

import PyQuante
import pyquante2


def pyquante_2_to_1_molecule(old_molecule):

    from PyQuante.Atom import Atom
    from PyQuante import Molecule

    # new_atoms = []
    # for old_atom in old_molecule.atoms:
    #     new_atom = Atom(atno=old_atom.Z,
    #                     x=old_atom.r[0],
    #                     y=old_atom.r[1],
    #                     z=old_atom.r[2])
    #     new_atoms.append(new_atom)

    # new_molecule = Molecule(name=old_molecule.name,
    #                         atomlist=new_atoms,
    #                         units=old_molecule.units,
    #                         charge=old_molecule.charge,
    #                         multiplicity=old_molecule.multiplicity)

    new_molecule = Molecule(name=old_molecule.name,
                            units=old_molecule.units,
                            charge=old_molecule.charge,
                            multiplicity=old_molecule.multiplicity)

    for old_atom in old_molecule.atoms:
        new_atom = Atom(atno=old_atom.Z,
                        x=old_atom.r[0],
                        y=old_atom.r[1],
                        z=old_atom.r[2])
        new_molecule.add_atom(new_atom)

    return new_molecule


def pyquante_2_to_1_basis(old_basis):

    from PyQuante import CGBF

    new_basis = []

    for old_cgbf in old_basis:
        # new_cgbf.norm = old_cgbf.norm
        # conversion from prims?
        # new_cgbf.pnorms = old_cgbf.pnorms
        # new_cgbf.pexps = old_cgbf.pexps
        # new_cgbf.pcoefs = old_cgbf.pcoefs
        new_cgbf = CGBF(old_cgbf.origin, old_cgbf.powers)
        for ccoef, old_pgbf in old_cgbf:
            # new_pgbf = PGBF(exponent=old_pgbf.exponent,
            #                 origin=old_pgbf.origin,
            #                 powers=old_pgbf.powers,
            #                 norm=old_pgbf.norm)
            new_cgbf.add_primitive(exponent=old_pgbf.exponent,
                                   coefficient=ccoef)
        new_basis.append(new_cgbf)

    return new_basis


def make_test_integrals_pyquante(args, mol, mol_basis):
    """A driver for making integrals for test comparisons using PyQuante 1
    and 2.
    """

    solver = pyquante2.uhf(mol, mol_basis)
    solver.converge(tol=1e-10, maxiters=1000)

    i1 = solver.i1

    C = solver.orbsa
    D = np.dot(C[:, :mol.nocc()], C[:, :mol.nocc()].transpose())

    # Save the one-electron integral matrices to disk.
    np.savetxt('pyquante2.S.txt', i1.S)
    np.savetxt('pyquante2.T.txt', i1.T)
    np.savetxt('pyquante2.V.txt', i1.V)
    np.savetxt('pyquante2.D.txt', D)

    # Save the energies to disk.
    np.savetxt('pyquante2.total_energy.txt', np.array([solver.energy]))
    np.savetxt('pyquante2.nuclear_repulsion_energy.txt', np.array([mol.nuclear_repulsion()]))

    # That was all pyquante2 stuff. Now for PyQuante!

    mol1 = pyquante_2_to_1_molecule(mol)
    bfs1 = pyquante_2_to_1_basis(mol_basis.bfs)

    from PyQuante.Ints import getFC

    for iat, at in enumerate(mol1.atoms, 1):
        FC_atom = getFC(bfs1, at.r)
        np.savetxt('PyQuante.FC{}.txt'.format(iat), FC_atom)
