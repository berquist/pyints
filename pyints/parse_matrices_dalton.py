#!/usr/bin/env python

"""parse_matrices_dalton.py: Parse matrices from DALTON output files
and save them to disk either as plain text or NumPy binary files.
"""

from __future__ import print_function
from __future__ import division

import re
import numpy as np


def getargs():
    """Get and parse command-line arguments."""

    import argparse

    parser = argparse.ArgumentParser()
    arg = parser.add_argument

    arg("outputfilename",
        help="""Name of the DALTON output file to parse for matrices.""")
    arg("--print-stdout",
        action="store_true",
        help="""Should all the parsed matrices be printed to stdout (usually
        the screen)?""")

    args = parser.parse_args()

    return args


# RE_ELEMENT = re.compile(r'(([1-9][0-9]*\.?[0-9]*)|(\.[0-9]+))([EeDd]?[+-]?[0-9]+)?')
# RE_EXPT = re.compile(r'[+-]?\d*$')
RE_ELEMENT = re.compile(r'([+-]?[0-9]*\.?[0-9]*|[+-]?\.[0-9]+)[EeDd]?([+-]?[0-9]+)?')
NBASIS_RE_STRING = r'total:\s+([0-9]+)\s+([-+]?\d*[.]?\d+)\s+([0-9]+)\s+([0-9]+)'


def parse_element_dalton(element):
    """Given a number that might appear in a DALTON output, especially one
    printed in a matrix, convert it to a float.

    The 'expt' regex specifically captures the exponent for when a 'D'
    isn't present in the number.

    Parameters
    ----------
    element : str

    Returns
    -------
    float
    """

    # res = RE_ELEMENT.findall(element)[0][0]
    # expt = RE_EXPT.findall(element)[0][0]

    # return float(res + 'e' + expt)

    if any(x in element for x in ['e', 'E', 'd', 'D']):
        match = [(p, q) for (p, q) in RE_ELEMENT.findall(element)
                 if p is not '' if q is not ''][0]
        return float('e'.join(match))
    else:
        return float(element)


def parse_matrix_dalton(outputfile):
    """Parse a matrix from a DALTON output file. `outputfile` is the file
    object being read from.

    Since DALTON doesn't print every line or column if all its
    elements fall below a certain threshold (is there an option for
    this?), we will parse and form a "sparse" (dictionary of
    dictionaries) representation, and separately/later convert this
    sparse representation to dense array format.

    ...which is why there's no matrix passed as an argument.

    Parameters
    ----------
    outputfile : file handle
        The file handle should be at a position near the desired matrix.

    Returns
    -------
    dict
        A sparse matrix of floats packed into a 2D dictionary, where each
        keys are single indices into non-zero matrix elements.
    """

    spmat = dict()

    line = ''
    while 'Column' not in line:
        line = next(outputfile)
    while '==== End of matrix output ====' not in line:
        chomp = line.split()
        if chomp == []:
            pass
        elif chomp[0] == 'Column':
            colindices = list(map(int, chomp[1::2]))
        else:
            rowindex = int(chomp[0])
            if rowindex not in spmat:
                spmat[rowindex] = dict()
            for colindex, element in zip(colindices, chomp[1:]):
                spmat[rowindex][colindex] = parse_element_dalton(element)
        line = next(outputfile)
    return spmat


def sparse_to_dense_matrix_dalton(spmat, dim):
    """Convert a "sparse" DALTON matrix to its full representation.

    TODO FIXME this almost certainly doesn't work properly for
    antisymmetric matrices or matrices without symmetry.

    Parameters
    ----------
    spmat : dictionary of dictionaries of floats
    dim : int

    Returns
    -------
    np.ndarray
    """
    densemat = np.zeros(shape=(dim, dim))
    for ridx in spmat:
        for cidx in spmat[ridx]:
            densemat[ridx-1][cidx-1] \
                = densemat[cidx-1][ridx-1] \
                = spmat[ridx][cidx]
    return densemat


# pylint: disable=invalid-name
def parse_spin_orbit_2el(outputfile_reversed, dim):
    """Read two-electron spin-orbit integrals from a DALTON output file.

    Parameters
    ----------
    outputfile_reversed : file handle
        The output file is iterative over in reverse order for ease of
        terminating parsing.
    dim : int

    Returns
    -------
    tuple of 3 4-dimensional np.ndarrays, one for each Cartesian
    component
    """
    x2spnorb = np.zeros(shape=(dim, dim, dim, dim))
    y2spnorb = np.zeros(shape=(dim, dim, dim, dim))
    z2spnorb = np.zeros(shape=(dim, dim, dim, dim))
    line = ''
    while '##' not in line:
        line = next(outputfile_reversed)
    while '##' in line:
        tokens = line.split()
        mu, nu, lm, sg = list(map(int, tokens[1:5]))
        component = tokens[5][1]
        element = parse_element_dalton(tokens[7])
        if component == 'X':
            x2spnorb[mu-1, nu-1, lm-1, sg-1] = element
        elif component == 'Y':
            y2spnorb[mu-1, nu-1, lm-1, sg-1] = element
        elif component == 'Z':
            z2spnorb[mu-1, nu-1, lm-1, sg-1] = element
        else:
            raise ValueError("Shouldn't be here...")
        line = next(outputfile_reversed)
    return x2spnorb, y2spnorb, z2spnorb


# We could parse these, but it's pointless since we don't have
# anything to compare them to from any other program package.
IGNORE_THESE_MATRICES = (
    'HUCKOVLP',
    'HUCKEL',
)

def parse_matrices_dalton(outputfilename, dump=False, ignore_headers=IGNORE_THESE_MATRICES):
    """The main routine. Finds all possible integral matrices in a DALTON
    outputfile automatically and parses them.
    """

    import os.path

    stub = os.path.splitext(outputfilename)[0]

    # Find all the matrix headers in an output file and turn them into
    # reasonable variable and filenames.
    matrix_headers_unmodified = []
    matrix_headers_varnames = []
    matrix_headers_filenames = []
    matchline = 'Integrals of operator: '
    with open(outputfilename) as outputfile:
        for line in outputfile:
            if matchline in line:
                if not any(ignore in line for ignore in ignore_headers):
                    # need to handle headers with spaces/multiple parts
                    start = line.index(matchline) + len(matchline)
                    matname = line[start:][:-2].strip()
                    varname = ''.join(matname.split()).replace(',', '_').replace('(', '_').replace(')', '_').replace('.', '_').replace('-', '_').lower()
                    filename = '.'.join([stub, 'integrals_AO_' + varname, 'txt'])
                    matrix_headers_unmodified.append(matname)
                    matrix_headers_varnames.append(varname)
                    matrix_headers_filenames.append(filename)

    # Determine the size of the basis. Assume N_AO == N_MO!
    with open(outputfilename) as outputfile:
        for line in outputfile:
            match = re.search(NBASIS_RE_STRING, line)
            if match:
                nbasis = int(match.groups()[-1])

    # Parse the output file for each matrix.
    parse_matrix_string = '{matrix_name} = sparse_to_dense_matrix_dalton(parse_matrix_dalton(outputfile), nbasis)'
    save_matrix_string = 'np.savetxt("{file_name}", {matrix_name})'
    matrices = dict()
    for i, matrix_name in enumerate(matrix_headers_unmodified):
        with open(outputfilename) as outputfile:
            for line in outputfile:
                if 'Integrals of operator: {}'.format(matrix_name) in line:
                    matrices[matrix_headers_varnames[i]] = sparse_to_dense_matrix_dalton(parse_matrix_dalton(outputfile), nbasis)
                    if dump:
                        np.savetxt(matrix_headers_filenames[i], matrices[matrix_headers_varnames[i]])
                    break

    # If present, dump the two-electron spin-orbit integrals.
    with open(outputfilename) as outputfile:
        outputfile_reversed = reversed(outputfile.readlines())
    for line in outputfile_reversed:
        if 'Number of written 2-el. spin-orbit integrals' in line:
            x2spnorb, y2spnorb, z2spnorb = parse_spin_orbit_2el(outputfile_reversed, nbasis)
            if dump:
                np.save('.'.join([stub, 'integrals_AO_x2spnorb.npy']), x2spnorb)
                np.save('.'.join([stub, 'integrals_AO_y2spnorb.npy']), y2spnorb)
                np.save('.'.join([stub, 'integrals_AO_z2spnorb.npy']), z2spnorb)
            break

    # Print all the parsed matrices to stdout if asked.
    if args.print_stdout:
        for matrix_name in matrix_headers_varnames:
            print(matrix_name)
            print(matrices[matrix_name])

    # Find the total energy and the nuclear repulsion energy.
    with open(outputfilename) as outputfile:
        for line in outputfile:
            if '@    Nuclear repulsion:' in line:
                V_nn = parse_element_dalton(line.split()[3])
            if '@     Converged SCF energy' in line:
                E_total = parse_element_dalton(line.split()[-2])

    if dump:
        np.savetxt('.'.join([stub, 'nuclear_repulsion_energy.txt']), np.array([V_nn]))
        np.savetxt('.'.join([stub, 'total_energy.txt']), np.array([E_total]))

    return locals()


if __name__ == '__main__':
    args = getargs()
    _locals = parse_matrices_dalton(args.outputfilename, dump=True)
