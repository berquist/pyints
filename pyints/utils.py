from functools import reduce
from itertools import combinations_with_replacement

import numpy as np


def fact2(n):
    return reduce(int.__mul__, range(n, 0, -2), 1)


def is_symmetric(arr):
    return (arr.transpose() == arr).all()


def antisymmetrize_L(mat):
    return np.triu(mat) - np.tril(mat) - np.diag(np.diag(mat))


def antisymmetrize_U(mat):
    return np.tril(mat) - np.triu(mat) - np.diag(np.diag(mat))


def pairs(it):
    return combinations_with_replacement(it, 2)


def iterator4(nbfs):
    for i, j in pairs(range(nbfs)):
        ij = (i*(i+1)/2) + j
        for k, l in pairs(range(nbfs)):
            kl = (k*(k+1)/2) + l
            if ij <= kl:
                yield i, j, k, l
    return
