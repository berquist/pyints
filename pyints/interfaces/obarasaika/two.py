import numpy as np

import obarasaika.obara_saika as os

from ...utils import iterator4


### two-electron repulsion integrals (ERI)

def coulomb_repulsion(za, la, ra, zb, lb, rb, zc, lc, rc, zd, ld, rd):
    return os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, la + lb + lc + ld)


def ERI(a, b, c, d):
    if d.contracted:
        return sum(cd * ERI(pd, c, a, b) for (cd, pd) in d)
    return a.norm * b.norm * c.norm * d.norm * \
        coulomb_repulsion(a.exponent, list(a.powers), a.origin,
                          b.exponent, list(b.powers), b.origin,
                          c.exponent, list(c.powers), c.origin,
                          d.exponent, list(d.powers), d.origin)


def makeERI(bfs):
    nbfs = len(bfs)
    ints = np.zeros(shape=(nbfs, nbfs, nbfs, nbfs))
    for i, j, k, l in iterator4(nbfs):
        ints[i, j, k, l] \
            = ints[i, j, l, k] \
            = ints[j, i, k, l] \
            = ints[j, i, l, k] \
            = ints[k, l, i, j] \
            = ints[k, l, j, i] \
            = ints[l, k, i, j] \
            = ints[l, k, j, i] \
            = ERI(bfs[i], bfs[j], bfs[k], bfs[l])
    return ints


### two-electron spin-orbit integrals (J2) from King and Furlani's
### formulation.

def spin_orbit_2_KF(za, zb, zc, zd, ra, rb, rc, rd, la, lb, lc, ld, component):
    ai, aj = za, zb
    # ak, al = zc, zd
    xi, yi, zi = la[0], la[1], la[2]
    xj, yj, zj = lb[0], lb[1], lb[2]
    xk, yk, zk = lc[0], lc[1], lc[2]
    xl, yl, zl = ld[0], ld[1], ld[2]
    if component == 2:
        return (   xi*yi*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi-1, yi, zi, xj, yj-1, zj, xk, yk, zk, xl, yl, zl]) \
                -2*ai*yj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi+1, yi, zi, xj, yj-1, zj, xk, yk, zk, xl, yl, zl]) \
                -2*aj*xi*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi-1, yi, zi, xj, yj+1, zj, xk, yk, zk, xl, yl, zl]) \
                +4*ai*aj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi+1, yi, zi, xj, yj+1, zj, xk, yk, zk, xl, yl, zl]) \
                  -yi*xj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi-1, zi, xj-1, yj, zj, xk, yk, zk, xl, yl, zl]) \
                +2*ai*xj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi+1, zi, xj-1, yj, zj, xk, yk, zk, xl, yl, zl]) \
                +2*aj*yi*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi-1, zi, xj+1, yj, zj, xk, yk, zk, xl, yl, zl]) \
                -4*ai*aj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi+1, zi, xj+1, yj, zj, xk, yk, zk, xl, yl, zl]))
    if component == 0:
        return (   yi*zj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi-1, zi, xj, yj, zj-1, xk, yk, zk, xl, yl, zl]) \
                -2*aj*yi*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi-1, zi, xj, yj, zj+1, xk, yk, zk, xl, yl, zl]) \
                -2*ai*zj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi+1, zi, xj, yj, zj-1, xk, yk, zk, xl, yl, zl]) \
                +4*ai*aj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi+1, zi, xj, yj, zj+1, xk, yk, zk, xl, yl, zl]) \
                  -zi*yj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi, zi-1, xj, yj-1, zj, xk, yk, zk, xl, yl, zl]) \
                +2*aj*zi*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi, zi-1, xj, yj+1, zj, xk, yk, zk, xl, yl, zl]) \
                +2*ai*yj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi, zi+1, xj, yj-1, zj, xk, yk, zk, xl, yl, zl]) \
                -4*ai*aj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi, zi+1, xj, yj+1, zj, xk, yk, zk, xl, yl, zl]))
    if component == 1:
        return (   zi*xj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi, zi-1, xj-1, yj, zj, xk, yk, zk, xl, yl, zl]) \
                -2*aj*zi*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi, zi-1, xj+1, yj, zj, xk, yk, zk, xl, yl, zl]) \
                -2*ai*xj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi, zi+1, xj-1, yj, zj, xk, yk, zk, xl, yl, zl]) \
                +4*ai*aj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi, yi, zi+1, xj+1, yj, zj, xk, yk, zk, xl, yl, zl]) \
                  -xi*zj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi-1, yi, zi, xj, yj, zj-1, xk, yk, zk, xl, yl, zl]) \
                +2*aj*xi*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi-1, yi, zi, xj, yj, zj+1, xk, yk, zk, xl, yl, zl]) \
                +2*ai*zj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi+1, yi, zi, xj, yj, zj-1, xk, yk, zk, xl, yl, zl]) \
                -4*ai*aj*os.get_coulomb(za, zb, zc, zd, ra, rb, rc, rd, [xi+1, yi, zi, xj, yj, zj+1, xk, yk, zk, xl, yl, zl]))


def J2_KF(a, b, c, d, component):
    if a.contracted:
        return sum(ca * J2_KF(pa, b, c, d, component) for (ca, pa) in a)
    if b.contracted:
        return sum(cb * J2_KF(a, pb, c, d, component) for (cb, pb) in b)
    if c.contracted:
        return sum(cc * J2_KF(a, b, pc, d, component) for (cc, pc) in c)
    if d.contracted:
        return sum(cd * J2_KF(a, b, c, pd, component) for (cd, pd) in d)
    return a.norm * b.norm * c.norm * d.norm * \
        spin_orbit_2_KF(a.exponent, b.exponent, c.exponent, d.exponent,
                        a.origin, b.origin, c.origin, d.origin,
                        list(a.powers), list(b.powers), list(c.powers), list(d.powers),
                        component)


def makeJ2_KF(bfs):
    nbfs = len(bfs)
    ints_X = np.zeros(shape=(nbfs, nbfs, nbfs, nbfs))
    ints_Y = np.zeros(shape=(nbfs, nbfs, nbfs, nbfs))
    ints_Z = np.zeros(shape=(nbfs, nbfs, nbfs, nbfs))
    for mu, a in enumerate(bfs):
        for nu, b in enumerate(bfs):
            for lm, c in enumerate(bfs):
                for sg, d in enumerate(bfs):
                    ints_X[mu, nu, lm, sg] = J2_KF(a, b, c, d, 0)
                    ints_Y[mu, nu, lm, sg] = J2_KF(a, b, c, d, 1)
                    ints_Z[mu, nu, lm, sg] = J2_KF(a, b, c, d, 2)
    return ints_X, ints_Y, ints_Z
