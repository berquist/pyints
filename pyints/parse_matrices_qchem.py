#!/usr/bin/env python

"""parse_matrices_qchem.py: Parse matrices from Q-Chem output files
and save them to disk either as plain text or NumPy binary files.
"""

from __future__ import print_function
from __future__ import division

import os
import re

import numpy as np

from .utils import print_matrices, dump_matrices


def qchem_parse_matrix(outputfile, matrix):
    """Parse a matrix from a Q-Chem output file. `outputfile` is the file
    object being read from. `matrix` is the NumPy array that we modify
    in-place.
    """
    # the shape of the matrix is the same as the dimension of the basis
    dim = matrix.shape[0]
    # The first line will always tell us the maximum number of columns wide
    # a block will be.
    line = next(outputfile)
    ncols = int(line.split()[-1])
    colcounter = 0
    while colcounter < dim:
        if line[:5].strip() == '':
            line = next(outputfile)
        rowcounter = 0
        while rowcounter < dim:
            row = list(map(float, line.split()[1:]))
            matrix[rowcounter][colcounter:colcounter+ncols] = row
            line = next(outputfile)
            rowcounter += 1
        colcounter += ncols
    return matrix


# These are the default matrices printed when calling
# `OneEMtrx().dump()`, plus a few I added in...:)
_MATRIX_HEADERS = (
    # 'Final Alpha MO Coefficients',
    'Final Alpha density matrix.',
    'Final Alpha Fock Matrix',
    'Core Hamiltonian Matrix',
    'Orthonormalization Matrix',
    'Overlap Matrix',
    'Kinetic Energy Matrix',
    'Nuclear Attraction Matrix',
    'Multipole Matrix (1,0,0)',
    'Multipole Matrix (0,1,0)',
    'Multipole Matrix (0,0,1)',
    'Multipole Matrix (2,0,0)',
    'Multipole Matrix (1,1,0)',
    'Multipole Matrix (0,2,0)',
    'Multipole Matrix (1,0,1)',
    'Multipole Matrix (0,1,1)',
    'Multipole Matrix (0,0,2)',
    'Angular Momentum Matrix (X-component)',
    'Angular Momentum Matrix (Y-component)',
    'Angular Momentum Matrix (Z-component)',
    'Spin-Orbit Interaction Matrix (X-component)',
    'Spin-Orbit Interaction Matrix (Y-component)',
    'Spin-Orbit Interaction Matrix (Z-component)'
)


NBASIS_RE_STRING = r'There are(\s+[0-9]+) shells and(\s+[0-9]+) basis functions'


def _varnameify(matname):
    """Turn a Q-Chem matrix header into an appropriate variable name."""
    varname = '_'.join(matname.split())
    varname = varname.replace(',', '').replace('(', '').replace(')', '').replace('.', '').replace('-', '_').lower()
    return varname


def _filenameify(stub, varname):
    """Turn a Q-Chem matrix variable name into an appropriate filename."""
    return '.'.join([stub, varname, 'txt'])


def parse_matrices_qchem(outputfilename, _matrix_headers=_MATRIX_HEADERS):
    """The main routine. Finds matrices with the given headers and parses
    them."""

    stub = os.path.splitext(os.path.basename(outputfilename))[0]

    # Turn the matrix headers into reasonable variable and filenames.
    matrix_headers = dict()
    matrix_filenames = dict()
    for matrix_name in _matrix_headers:
        varname = _varnameify(matrix_name)
        filename = _filenameify(stub, varname)
        matrix_headers[varname] = matrix_name
        matrix_filenames[varname] = filename

    # Determine the size of the basis. Assume N_AO == N_MO!
    with open(outputfilename) as outputfile:
        for line in outputfile:
            match = re.findall(NBASIS_RE_STRING, line)
            if match:
                nbasis = int(match[0][1].strip())

    # Parse the output file for each matrix.
    for matrix_name in matrix_headers:
        with open(outputfilename) as outputfile:
            for line in outputfile:
                if matrix_name in line:
                    matrices[matrix_name] = np.ones(shape=(nbasis, nbasis)) * -99
                    matrices[matrix_name] = qchem_parse_matrix(outputfile, matrices[matrix_name])
                    break

    # Find the total energy and the nuclear repulsion energy.
    # with open(outputfilename) as outputfile:
    #     for line in outputfile:
    #         if 'Nuclear Repulsion Energy' in line:
    #             V_nn = float(line.split()[4])
    #         if 'Total energy in the final basis set' in line:
    #             E_total = float(line.split()[-1])

    # np.savetxt('qchem.nuclear_repulsion_energy.txt', np.array([V_nn]))
    # np.savetxt('qchem.total_energy.txt', np.array([E_total]))

    return matrix_headers, matrix_filenames, matrices


if __name__ == '__main__':

    def getargs():
        """Get and parse command-line arguments."""

        import argparse

        parser = argparse.ArgumentParser()
        arg = parser.add_argument

        arg("outputfilename",
            help="""Name of the Q-Chem output file to parse for matrices.""")
        arg("--print-stdout",
            action="store_true",
            help="""Should all the parsed matrices be printed to stdout (usually
            the screen)?""")

        args = parser.parse_args()

        return args

    args = getargs()
    for outputfilename in args.outputfilename:
        matrix_headers, matrix_filenames, matrices = parse_matrices_qchem(outputfilename)
        if args.print_stdout:
            print('=' * 70)
            print(stub)
            print_matrices(matrix_headers, matrices)
        stub = os.path.splitext(os.path.basename(outputfilename))[0]
        dump_matrices(stub, matrix_filenames, matrices)
