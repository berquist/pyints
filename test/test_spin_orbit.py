from __future__ import print_function

import itertools as i

from pyints.two import spin_orbit_2_KF


# def test_spin_orbit_2_KF():
#     za = 0.1688554
#     zb = 0.0480887
#     zc = zb
#     zd = zb

#     ra = [0., 0., 3.01392398]
#     rb = [0.0, 0.0, 1.0]
#     rc = [0.0, 0.0, 2.0]
#     rd = [0.0, 0.0, 4.0]

#     la = [2, 0, 0]
#     lb = [2, 0, 0]
#     lc = [2, 0, 0]
#     ld = [1, 1, 0]

#     component = 2

#     for ta, tb, tc, td in set(i.permutations((tuple(la), tuple(lb), tuple(lc), tuple(ld)), 4)):
#         print(ta, tb, tc, td,
#               spin_orbit_2_KF(za, zb, zc, zd, ra, rb, rc, rd, list(ta), list(tb), list(tc), list(td), component))

if __name__ == '__main__':
    pass
