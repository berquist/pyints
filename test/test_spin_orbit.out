(2, 0, 0) (2, 0, 0) (2, 0, 0) (1, 1, 0) 29743.2645024087
(1, 1, 0) (2, 0, 0) (2, 0, 0) (2, 0, 0) -20006.2186467759
(2, 0, 0) (1, 1, 0) (2, 0, 0) (2, 0, 0) -633556.745534277
(2, 0, 0) (2, 0, 0) (1, 1, 0) (2, 0, 0) 29743.2645024087
